### PR TITLE
crystal_1_2: add support for aarch64-darwin

### PR DESCRIPTION
**Pull request by winterqt**
###### Motivation for this change

Crystal 1.2 added support for aarch64-darwin.

cc @NixOS/crystal-lang

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
Result of `nixpkgs-review pr 154194` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>invidious</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>crystal (crystal_1_2)</li>
    <li>crystal2nix</li>
    <li>icr</li>
    <li>lucky-cli</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_16)</li>
    <li>shards_0_15</li>
    <li>thicket</li>
  </ul>
</details>

File: pkgs/development/compilers/crystal/default.nix [hunk 1]
```diff
@@ -238,10 +250,8 @@ rec {
   crystal_1_2 = generic {
     version = "1.2.2";
     sha256 = "sha256-nyOXhsutVBRdtJlJHe2dALl//BUXD1JeeQPgHU4SwiU=";
-    binary = crystal_1_1;
+    binary = if isAarch64Darwin then binaryCrystal_1_2 else crystal_1_1;
   };
 
   crystal = crystal_1_2;
-
-  crystal2nix = callPackage ./crystal2nix.nix { };
```
Comment (fabianhjr) [inline pkgs/development/compilers/crystal/default.nix:246]: Is this removal intentional? Seems unrelated

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -11729,7 +11729,7 @@ with pkgs;
   copper = callPackage ../development/compilers/copper {};
 
   inherit (callPackages ../development/compilers/crystal {
-    llvmPackages = llvmPackages_10;
+    llvmPackages = if stdenv.system == "aarch64-darwin" then llvmPackages else llvmPackages_10;
```
Comment (fabianhjr) [inline pkgs/top-level/all-packages.nix:11732]: Is llvmPackages ahead of llvmPackags_10?

File: pkgs/development/compilers/crystal/default.nix (hunk 1 above)
Comment (winterqt) [reply pkgs/development/compilers/crystal/default.nix:246]: It's intentional, since it's wrong and I assume it was just never removed when the file it's pointing to was. Do you want me to split it into a separate commit or PR?

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (winterqt) [reply pkgs/top-level/all-packages.nix:11732]: It is ahead of llvmPackages_10, yes. (llvmPackages_10 and below are broken on aarch64-darwin.)

What would be the best version to pin it to, in your opinion? I'm guessing llvmPackages_11 would be the best since it's the closest version to llvmPackages_10. (Why do we pin to llvmPackages_10, anyways? Does upstream recommend/only support some versions or something?)

File: pkgs/development/compilers/crystal/default.nix (hunk 1 above)
Comment (peterhoeg) [reply pkgs/development/compilers/crystal/default.nix:246]: Separate commit would be great, thanks

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (peterhoeg) [reply pkgs/top-level/all-packages.nix:11732]: We've had some issues in the past with specific llvm versions as we were typically ahead of what crystal upstream would support.

**Review by fabianhjr — APPROVED** [review]
diff lgtm but can't test (no darwin nor aarch64 hardware)

File: pkgs/development/compilers/crystal/default.nix [hunk 3]
```diff
@@ -53,6 +55,8 @@ let
         tar --strip-components=1 -C $out -xf ${src}
         patchShebangs $out/bin/crystal
       '';
+
+      broken = lib.versionOlder version "1.2.0" && isAarch64Darwin;
```
Comment (cole-h) [inline pkgs/development/compilers/crystal/default.nix:59]: I don't think that's where `broken` goes -- 99% sure it only does things when inside the `meta` attribute, but I could be wrong.

Comment (winterqt) [reply pkgs/development/compilers/crystal/default.nix]: Good catch, thanks.

**Comment by fabianhjr** [comment]
Hi @tomberek, diff lgtm but cannot test since I have neither ARMv8/AArch64 hardware nor a macOS license/image.
